### PR TITLE
Bug 1813913 - Added support for Custom Tabs setColorScheme & setColorSchemeParams API

### PR DESCRIPTION
**Pull request by t-p-white**
ℹ️ to test this locally, recommend using the Custom Tabs demo and manually modifying the `CustomUIActivity`  if required.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1813913

### CODE REVIEW

File: android-components/components/support/utils/src/main/java/mozilla/components/support/utils/ColorUtils.kt [hunk 1]
```diff
@@ -18,12 +19,12 @@ object ColorUtils {
     }
 
     /**
-     * Returns true if the color is dark enough that white text should be used on top of it.
+     * @return true if the given [color] is dark enough that white text should be used on top of it.
      */
     @JvmStatic
     @SuppressWarnings("MagicNumber")
     fun isDark(@ColorInt color: Int): Boolean {
-        if (color == Color.WHITE || Color.alpha(color) < 128) {
```
Comment (t-p-white) [inline android-components/components/support/utils/src/main/java/mozilla/components/support/utils/ColorUtils.kt:26]: This method was leading to unexpected results with various colours, notably using green. Updated to use `calculateLuminance` which does this calculations for us

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/CustomTabConfig.kt [hunk 2]
```diff
@@ -9,34 +9,41 @@ import android.graphics.Bitmap
...
- * @property sessionToken The token associated with the custom tab.
+ * @property sessionToken Optional token associated with the custom tab.
  * @property externalAppType How this custom tab is being displayed.
  */
 data class CustomTabConfig(
-    @ColorInt val toolbarColor: Int? = null,
+    @ColorScheme val colorScheme: Int? = null,
+    val defaultColorSchemeParams: ColorSchemeParams? = null,
+    val lightColorSchemeParams: ColorSchemeParams? = null,
+    val darkColorSchemeParams: ColorSchemeParams? = null,
```
Comment (Amejia481) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/CustomTabConfig.kt:39]: Maybe we could have the parameters as part of data classes ? Something like a theme class or similar just to group all of the as parameter so we don't group the list parameters for creating a `CustomTabConfig`, what do you think?

Comment (Amejia481) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/CustomTabConfig.kt:39]: Similar to what we did for `ColorSchemeParams`.

Comment (t-p-white) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/CustomTabConfig.kt:39]: something like this?
```kotlin
data class CustomTabThemes(
    val defaultColorSchemeParams: ColorSchemeParams? = null,
    val lightColorSchemeParams: ColorSchemeParams? = null,
    val darkColorSchemeParams: ColorSchemeParams? = null,
)
```

File: android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Window.kt [hunk 3]
```diff
@@ -9,30 +9,29 @@ import android.os.Build.VERSION.SDK_INT
...
 import androidx.annotation.ColorInt
 import androidx.core.view.WindowInsetsControllerCompat
-import mozilla.components.support.utils.ColorUtils.isDark
 
 /**
  * Colors the status bar.
  * If the color is light enough, a light status bar with dark icons will be used.
  */
 fun Window.setStatusBarTheme(@ColorInt toolbarColor: Int) {
-    createWindowInsetsController().isAppearanceLightStatusBars =
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Window.kt:19]: I don't believe this is required

File: android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Window.kt [hunk 4]
```diff
@@ -9,30 +9,29 @@ import android.os.Build.VERSION.SDK_INT
...
     statusBarColor = toolbarColor
 }
 
 /**
- * Colors the navigation bar.
- * If the color is light enough, a light navigation bar with dark icons will be used.
+ * Set the navigation bar background and divider colors.
  */
-fun Window.setNavigationBarTheme(@ColorInt toolbarColor: Int) {
-    createWindowInsetsController().isAppearanceLightNavigationBars =
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Window.kt:29]: I don't believe this is required

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/CustomTabConfig.kt (hunk 2 above)
Comment (Amejia481) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/state/CustomTabConfig.kt:39]: Yes

File: android-components/components/support/utils/src/main/java/mozilla/components/support/utils/ColorUtils.kt (hunk 1 above)
Comment (Amejia481) [reply android-components/components/support/utils/src/main/java/mozilla/components/support/utils/ColorUtils.kt:26]: ✨

File: android-components/components/feature/customtabs/src/test/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeatureTest.kt [hunk 5]
```diff
@@ -877,7 +887,7 @@ class CustomTabsToolbarFeatureTest {
         doNothing().`when`(pendingIntent).send(any(), anyInt(), any())
 
         verify(pendingIntent).send(any(), anyInt(), intentCaptor.capture())
-        assertEquals("https://github.com/mozilla-mobile/android-components", intentCaptor.value.getDataString())
+        assertEquals("https://github.com/mozilla-mobile/android-components", intentCaptor.value.dataString)
```
Comment (t-p-white) [inline android-components/components/feature/customtabs/src/test/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeatureTest.kt:890]: house keeping

File: android-components/components/feature/customtabs/src/test/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeatureTest.kt [hunk 6]
```diff
@@ -569,7 +579,7 @@ class CustomTabsToolbarFeatureTest {
         button.performClick()
 
         verify(pendingIntent).send(any(), anyInt(), intentCaptor.capture())
-        assertEquals("https://github.com/mozilla-mobile/android-components", intentCaptor.value.getDataString())
+        assertEquals("https://github.com/mozilla-mobile/android-components", intentCaptor.value.dataString)
```
Comment (t-p-white) [inline android-components/components/feature/customtabs/src/test/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeatureTest.kt:582]: house keeping

**Review by Amejia481 — APPROVED** [review]
LGTM!

File: android-components/components/feature/customtabs/src/test/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeatureTest.kt (hunk 5 above)
Comment (Amejia481) [reply android-components/components/feature/customtabs/src/test/java/mozilla/components/feature/customtabs/CustomTabsToolbarFeatureTest.kt:890]: 🧹

**Comment by t-p-white** [comment]
Thanks for the review @Amejia481. Just FYI the latest push fixes:

1. lint failing due to formatting
2. The condition in `CustomTabsToolbarFeature` L205 for `window?.setNavigationBarTheme(navigationBarColor, navigationBarDividerColor)` should be an OR not the AND that I had originally

**Pull request merged**